server/commands/write.go

```diff
@@ -87,6 +113,72 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	return nil
 }
 
+// validateTypesForTuple makes sure that when writing a tuple, the types are compatible.
+// 1. If the tuple is of the form (person:bob, reader, doc:budget), then the type "doc", relation "reader" allows type "person".
+// 2. If the tuple is of the form (group:abc#member, reader, doc:budget), then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of the form (*, reader, doc:budget), we allow it only if the type "doc" relation "reader" allows at least one type (with no relation)
+func (c *WriteCommand) validateTypesForTuple(authModel *openfgapb.AuthorizationModel, schemaVersion typesystem.SchemaVersion, tk *openfgapb.TupleKey, dbCallsCounter utils.DBCallCounter) error {
+	objectType, _ := tupleUtils.SplitObject(tk.GetObject())    // e.g. "doc"
+	userType, userID := tupleUtils.SplitObject(tk.GetUser())   // e.g. (person, bob) or (group, abc#member) or ("", *)
+	_, userRel := tupleUtils.SplitObjectRelation(tk.GetUser()) // e.g. (person:bob, "") or (group:abc, member) or (*, "")
+
+	// find the type information
+	var typeDefinitionForObject *openfgapb.TypeDefinition
+	var typeDefinitionForUser *openfgapb.TypeDefinition
+	for _, td := range authModel.GetTypeDefinitions() {
```

Comment:
Please use the typesystem package here instead. You should be able to do:

```
relation, err := typesys.Relation(objectType, tk.GetRelation())
```
and then use the relation
```
relation.GetTypeInfo().GetDirectlyRelatedUserTypes()
// etc..
```